# Welcome to lab_🐼s!

Last week, you did amazing work setting up your computer to do data science!  Let's now use your computer to do data science for the first time!

A few tips to remember:

- **You are not alone on your journey in learning programming!**  You have your lab Teaching Assistant (TA), your Course Assistants (CAs), your lab group, and the professors (Prof. Karle and Prof. Wade), who are all here to help you out!
- If you find yourself stuck for more than a few minutes, ask a neighbor or course staff for help!  When you are giving help to your neighbor, explain the **idea and approach** to the problem without sharing the answer itself so they can have the same **<i>ah-hah</i>** moment!
- We are here to help you!  Don't feel embarrassed or shy to ask us for help!

Let's get started!

<hr style="color: #DD3403;">

# Part 0: Your Lab Group!

We find it's best when you're programming with a group of people who you can help out and can help you out!

- A key part of Data Science is being able to **communicate about your work**.  If you are the best programmer in the world, but no one uses your code, then the code you're writing is not getting run.  By learning strong communication skills, you will be able to **explain your code**, and the underlying data science concepts, to everyone -- data scientists and non-scientists alike!

- **Every lab** you will need to complete some parts of the lab **with your group**.  There will be both **"group discussions"** and **"individual reflections"** as part of every lab.

## Create Your Group

If you have not done so already, **RIGHT NOW** find *(ideally)* two other people to make a group of **two, three, or four** people.   *(We find groups of three work the best, but you can be a group of two or four -- no solo groups, and no groups of five!*)

- If you're at a group of five, split into a group of two and a group of three.  You can still help each other as a larger group, but the group questions should be answered in your smaller group of 2-4 people.

Once you have your group, **introduce yourselves to each other! :)**

## Record Your Group

Since you know Python, edit the following cell to add strings about your group:

In [ ]:
# First, meet your CAs and TA if you haven't already!
# ...first name is enough, we'll know who they are! :)
ta_name = "Ridha Alkhabaz"
ca1_name = "Jackson Clark"
ca2_name = "Qiran Hu"
ca3_name = ""


# As part of your introduction, learn your lab partner's name, major, and
# learn something new about them that you never knew!
# - Groups of 3 are ideal :)
# - However groups of 2 or 4 are fine too!
partner1_name = "david poderczerwinksi"
partner1_netid = "davidp7"
partner1_major = "business"

partner2_name = "Allie Bosch"
partner2_netid = "allieb12"
partner2_major = "information science"

partner3_name = ""
partner3_netid = ""
partner3_major = ""

<hr style="color: #DD3403;">

# Part 1: 🧑‍🤝‍🧑 Hello Dataset!

In lecture, we have already begun to explore the "Party Dataset" that you created.  We will continue to use the Party Dataset all semester long, but we want even more datasets -- and we want you to come up with the questions!

In this lab, **you** will come up with questions to ask the entire class for the **"Hello Dataset"**!

## Hello Dataset Overview

Over the weekend, your lab homework will be to fill out a class-wide "Hello Survey".  Right now, with your group, you will come up some of the questions that should be on this survey *(and you will do analysis on your questions next week!)*.

With your group, come up with four questions below!  Before you get started, a few notes:

- The survey will be **NOT be anonymous** and the results will be shared with the class to do analysis on, so you should **only ask questions you feel comfortable to share with the class**.
- **Questions that are inappropriate** for a classroom environment will be thrown out and you won't earn lab credit points for any question we throw out.  (Keep everything at most PG/PG-13.)
- You are going to submit **one set of questions for your whole group**.  You must decide on the questions together.

## Q1: Question with a Binary Response

A question with a **binary response** is a question that has only two answers.  For example:
- *Do you prefer cats or dogs?*
- *Were you born in Illinois?*

With your group (you must all decide on a single question together), come up with a question that you think will have **exactly 50%/50%** split between the two binary outcomes.  *(The question that gets closest to 50%/50% will get extra credit!)*

**What is your binary question?**

*(✏️ Even or Odd? ✏️)*
# even or odd

**Why do you think this is a good question?  Write at least two complete sentences on why you believe your question will have a 50%/50% split.**

#we think that the question would allow anyone to chose whatever they want. it also ensures that with a large sample the outcomes would be 50/50. 

## Q2: Question with a Numeric Response

A question with a **numeric response** requires a number for an answer.  For example:

- *What is your height?*
- *What day were you born?*

With your group (you must all decide on a single question together), come up with a question and a prediction on what you think the most frequent answer will be to your question.

**What is your numeric question?**

*(what month were you born in "01, 02, etc.())*

**What do you think the most frequent answer will be?**

*(a borthday in the summer, May)*

**Why do you think this answer will be the most frequent answer?  Write at least one complete sentence on why you believe that it will be the most frequent answer.**

*(✏️ We think so because summer babies are more common in general than winter. there are also statistics that can back this up. ✏️)*

## Q3: Question with a (non-numeric) Textual Response

A question with a (non-numeric) **textual response** requires a word or two for an answer.  For example:

- *What is your favorite sport to play?*
- *What is the location of your favorite trip/vacation?*

With your group (you must all decide on a single question together), come up with a question where at least two members of your group share the same answer.

**What is your textual response question?**

*(✏️ What is your favorite place to study at on campus?✏️)*

**What is the shared answer between your group?  Do you also share this answer (if not, what's your answer to the question)?**

*(✏️ David says his dorm or ACES library ✏️)*

## Q4: Open Question

One final question -- no restrictions on the format!  Is there a great question your group discussed that you did not get to use one of the first three questions but you want to know the classes' answer?

**What is your final question for the whole class?**

*(✏️ What are your plans post graduation? ✏️)*

## Submit Your Questions to Wade and Karle

Your lab TAs will be grading your lab notebook for your full responses, but we want to make sure to get the survey with your questions out to everyone ASAP!

- Pick one member of your group as the "question submitter".
- If you're the question submitter, complete this Google Form for your group (requires @illinois.edu login): https://forms.gle/6HQ6MLyEhq7zK36i9
- Make sure to do this **during your lab section**.


<hr style="color: #DD3403;">

# Part 2: Importing the Course Catalog Dataset

In lecture, you learned we will use the `pandas` **library** for working with data.  Libraries are collections of functions and other objects that extend what Python can do by itself.  The `pandas` library is the most widely used library for data science.

To use the `pandas` library, we must **import** it into your notebook.  In every lab, you will need to import the library just like we did in lecture!

In the following cell, replace `...` with the code to import pandas as `pd`:

In [6]:
# Write the code to import pandas as pd:
import pandas as pd 

In lecture, you saw the "Course Catalog" dataset.  Using the `read_csv` function, load the course catalog dataset into a variable called `df`.  The URL for this dataset is:
```
https://waf.cs.illinois.edu/discovery/course-catalog.csv
```

Remember, URLs passed to `read_csv` are **string** variables, so remember to use **quotation marks**. 

In [4]:
# Load the course catalog dataset into a variable named `df`:
df = pd.read_csv("https://waf.cs.illinois.edu/discovery/course-catalog.csv") 
df

Year    Term YearTerm Subject  Number                          Name  \
0      2024  Spring  2024-sp     AAS     100  Intro Asian American Studies   
1      2024  Spring  2024-sp     AAS     100  Intro Asian American Studies   
2      2024  Spring  2024-sp     AAS     100  Intro Asian American Studies   
3      2024  Spring  2024-sp     AAS     100  Intro Asian American Studies   
4      2024  Spring  2024-sp     AAS     100  Intro Asian American Studies   
...     ...     ...      ...     ...     ...                           ...   
11898  2024  Spring  2024-sp    WRIT     303          Writing Across Media   
11899  2024  Spring  2024-sp    WRIT     303          Writing Across Media   
11900  2024  Spring  2024-sp    YDSH     102          Beginning Yiddish II   
11901  2024  Spring  2024-sp    YDSH     102          Beginning Yiddish II   
11902  2024  Spring  2024-sp    ZULU     202            Elementary Zulu II   

                                             Description Credit Hours  \
0      Interdisciplinary introduction to the basic co...     3 hours.   
1      Interdisciplinary introduction to the basic co...     3 hours.   
2      Interdisciplinary introduction to the basic co...     3 hours.   
3      Interdisciplinary introduction to the basic co...     3 hours.   
4      Interdisciplinary introduction to the basic co...     3 hours.   
...                                                  ...          ...   
11898                    Same as INFO 303. See INFO 303.     3 hours.   
11899                    Same as INFO 303. See INFO 303.     3 hours.   
11900  Continuation of YDSH 101 focusing on comprehen...     4 hours.   
11901  Continuation of YDSH 101 focusing on comprehen...     4 hours.   
11902  Continuation of ZULU 201 with introduction of ...     5 hours.   

                                            Section Info  \
0                                                    NaN   
1                                                    NaN   
2                                                    NaN   
3                                                    NaN   
4                                                    NaN   
...                                                  ...   
11898                    Same as INFO 303. See INFO 303.   
11899                    Same as INFO 303. See INFO 303.   
11900                                                NaN   
11901                                                NaN   
11902  Same as AFST 252. Participation in the languag...   

                                       Degree Attributes  ... Section Status  \
0      Social & Beh Sci - Soc Sci, and Cultural Studi...  ...              A   
1      Social & Beh Sci - Soc Sci, and Cultural Studi...  ...              A   
2      Social & Beh Sci - Soc Sci, and Cultural Studi...  ...              A   
3      Social & Beh Sci - Soc Sci, and Cultural Studi...  ...              A   
4      Social & Beh Sci - Soc Sci, and Cultural Studi...  ...              A   
...                                                  ...  ...            ...   
11898                       Advanced Composition course.  ...              A   
11899                       Advanced Composition course.  ...              A   
11900                                                NaN  ...              A   
11901                                                NaN  ...              A   
11902                                                NaN  ...              A   

                Enrollment Status                   Type Type Code Start Time  \
0                            Open  Discussion/Recitation       DIS   09:00 AM   
1                            Open  Discussion/Recitation       DIS   10:00 AM   
2                            Open  Discussion/Recitation       DIS   11:00 AM   
3                            Open  Discussion/Recitation       DIS   12:00 PM   
4                            Open  Discussion/Recitation       DIS   01:00 PM   
...                      

In [5]:
# Look at `df` to confirm that this worked:
df

Year    Term YearTerm Subject  Number                          Name  \
0      2024  Spring  2024-sp     AAS     100  Intro Asian American Studies   
1      2024  Spring  2024-sp     AAS     100  Intro Asian American Studies   
2      2024  Spring  2024-sp     AAS     100  Intro Asian American Studies   
3      2024  Spring  2024-sp     AAS     100  Intro Asian American Studies   
4      2024  Spring  2024-sp     AAS     100  Intro Asian American Studies   
...     ...     ...      ...     ...     ...                           ...   
11898  2024  Spring  2024-sp    WRIT     303          Writing Across Media   
11899  2024  Spring  2024-sp    WRIT     303          Writing Across Media   
11900  2024  Spring  2024-sp    YDSH     102          Beginning Yiddish II   
11901  2024  Spring  2024-sp    YDSH     102          Beginning Yiddish II   
11902  2024  Spring  2024-sp    ZULU     202            Elementary Zulu II   

                                             Description Credit Hours  \
0      Interdisciplinary introduction to the basic co...     3 hours.   
1      Interdisciplinary introduction to the basic co...     3 hours.   
2      Interdisciplinary introduction to the basic co...     3 hours.   
3      Interdisciplinary introduction to the basic co...     3 hours.   
4      Interdisciplinary introduction to the basic co...     3 hours.   
...                                                  ...          ...   
11898                    Same as INFO 303. See INFO 303.     3 hours.   
11899                    Same as INFO 303. See INFO 303.     3 hours.   
11900  Continuation of YDSH 101 focusing on comprehen...     4 hours.   
11901  Continuation of YDSH 101 focusing on comprehen...     4 hours.   
11902  Continuation of ZULU 201 with introduction of ...     5 hours.   

                                            Section Info  \
0                                                    NaN   
1                                                    NaN   
2                                                    NaN   
3                                                    NaN   
4                                                    NaN   
...                                                  ...   
11898                    Same as INFO 303. See INFO 303.   
11899                    Same as INFO 303. See INFO 303.   
11900                                                NaN   
11901                                                NaN   
11902  Same as AFST 252. Participation in the languag...   

                                       Degree Attributes  ... Section Status  \
0      Social & Beh Sci - Soc Sci, and Cultural Studi...  ...              A   
1      Social & Beh Sci - Soc Sci, and Cultural Studi...  ...              A   
2      Social & Beh Sci - Soc Sci, and Cultural Studi...  ...              A   
3      Social & Beh Sci - Soc Sci, and Cultural Studi...  ...              A   
4      Social & Beh Sci - Soc Sci, and Cultural Studi...  ...              A   
...                                                  ...  ...            ...   
11898                       Advanced Composition course.  ...              A   
11899                       Advanced Composition course.  ...              A   
11900                                                NaN  ...              A   
11901                                                NaN  ...              A   
11902                                                NaN  ...              A   

                Enrollment Status                   Type Type Code Start Time  \
0                            Open  Discussion/Recitation       DIS   09:00 AM   
1                            Open  Discussion/Recitation       DIS   10:00 AM   
2                            Open  Discussion/Recitation       DIS   11:00 AM   
3                            Open  Discussion/Recitation       DIS   12:00 PM   
4                            Open  Discussion/Recitation       DIS   01:00 PM   
...                      

## 🔬 Test Case Checkpoint 🔬

Throughout the lab, we will have "checkpoints" at key points.  Most checkpoints will include **test cases** that verify that your notebook appears to be correct.

- The test cases are **NOT** a guarantee that everything is perfect, but we try our very best to catch any common mistakes you might make!
- If you see `AssertionError` error message when running the test case, we provide a helpful hint on what may not be correct.
- If you see a party popper emoji and a message that all tests pass, you passed the checkpoint!
- **Checkpoints** are a good time to check with your group to make everyone gets to the checkpoint together!

In [7]:
## == CHECKPOINT TEST CASES for Part 2 ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.
assert( 'df' in vars() ), "Did you define a variable called df?"
assert( len(df) == 11903 ), "It looks like you did not import the correct dataset?"


## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


<hr style="color: #DD3403;">

# Part 3: STAT / CS / IS Majors

Data Science DISCOVERY is a unique course that is cross-listed across three departments: Statistics, Computer Science, and Information Sciences.  Let's explore courses from all three of those subjects!

### Puzzle 3.1

Select only the rows that have a `Subject` of `STAT`.  Assign this to the variable name `df_STAT` instead of the usual `df`.
- Make sure your result contains only statistics courses!

In [8]:
df_STAT = df[df.Subject == "STAT"]
df_STAT

Year    Term YearTerm Subject  Number                           Name  \
11168  2024  Spring  2024-sp    STAT     100                     Statistics   
11169  2024  Spring  2024-sp    STAT     100                     Statistics   
11170  2024  Spring  2024-sp    STAT     100                     Statistics   
11171  2024  Spring  2024-sp    STAT     107         Data Science Discovery   
11172  2024  Spring  2024-sp    STAT     107         Data Science Discovery   
...     ...     ...      ...     ...     ...                            ...   
11312  2024  Spring  2024-sp    STAT     587     Hierarchical Linear Models   
11313  2024  Spring  2024-sp    STAT     590  Individual Study and Research   
11314  2024  Spring  2024-sp    STAT     593                STAT Internship   
11315  2024  Spring  2024-sp    STAT     593                STAT Internship   
11316  2024  Spring  2024-sp    STAT     599                Thesis Research   

                                             Description    Credit Hours  \
11168  First course in probability and statistics at ...        3 hours.   
11169  First course in probability and statistics at ...        3 hours.   
11170  First course in probability and statistics at ...        3 hours.   
11171  Data Science Discovery is the intersection of ...        4 hours.   
11172  Data Science Discovery is the intersection of ...        4 hours.   
...                                                  ...             ...   
11312       Same as PSYC 587 and EPSY 587. See EPSY 587.        4 hours.   
11313  Directed reading and research. Approved for le...   0 TO 8 hours.   
11314  Supervised, off-campus experience in a field i...   0 TO 8 hours.   
11315  Supervised, off-campus experience in a field i...   0 TO 8 hours.   
11316  Approved for S/U grading only. May be repeated...  0 TO 16 hours.   

                                            Section Info  \
11168  Credit is not given for both STAT 100 and any ...   
11169  Credit is not given for both STAT 100 and any ...   
11170  Credit is not given for both STAT 100 and any ...   
11171                         Same as CS 107 and IS 107.   
11172                         Same as CS 107 and IS 107.   
...                                                  ...   
11312       Same as PSYC 587 and EPSY 587. See EPSY 587.   
11313  Approved for letter and S/U grading. May be re...   
11314  Approved for letter and S/U grading. Prerequis...   
11315  Approved for letter and S/U grading. Prerequis...   
11316  Approved for S/U grading only. May be repeated...   

                      Degree Attributes  ... Section Status  \
11168  Quantitative Reasoning I course.  ...              A   
11169  Quantitative Reasoning I course.  ...              A   
11170  Quantitative Reasoning I course.  ...              A   
11171  Quantitative Reasoning I course.  ...              A   
11172  Quantitative Reasoning I course.  ...              A   
...                                 ...  ...            ...   
11312                               NaN  ...              A   
11313                               NaN  ...              A   
11314                               NaN  ...              A   
11315                               NaN  ...              A   
11316                               NaN  ...              A   

                Enrollment Status                   Type Type Code Start Time  \
11168                        Open                Lecture       LEC   11:00 AM   
11169                      Closed                Lecture       LEC   12:00 PM   
11170           Open (Restricted)                Lecture       LEC   02:00 PM   
11171                      Closed                Lecture       LEC   10:00 AM   
11172                      Closed  Laboratory-Discussion       LBD   01:00 PM   
...                           ...                    ...       ...        ...   
11312  CrossListOpen (Restricted)                Lecture       LEC   10:00 AM   
11313              

### Puzzle 3.2

Select only the rows that have a `Subject` of `CS`.  Assign this to the variable name `df_CS` instead of the usual `df`.
- Make sure your result contains only computer science courses!

In [9]:
df_CS = df[df.Subject == "CS"]
df_CS

Year    Term YearTerm Subject  Number                              Name  \
3026  2024  Spring  2024-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3027  2024  Spring  2024-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3028  2024  Spring  2024-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3029  2024  Spring  2024-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3030  2024  Spring  2024-sp      CS     101  Intro Computing: Engrg &amp; Sci   
...    ...     ...      ...     ...     ...                               ...   
3402  2024  Spring  2024-sp      CS     598                    Special Topics   
3403  2024  Spring  2024-sp      CS     598                    Special Topics   
3404  2024  Spring  2024-sp      CS     598                    Special Topics   
3405  2024  Spring  2024-sp      CS     598                    Special Topics   
3406  2024  Spring  2024-sp      CS     599                   Thesis Research   

                                            Description    Credit Hours  \
3026  Fundamental principles, concepts, and methods ...        3 hours.   
3027  Fundamental principles, concepts, and methods ...        3 hours.   
3028  Fundamental principles, concepts, and methods ...        3 hours.   
3029  Fundamental principles, concepts, and methods ...        3 hours.   
3030  Fundamental principles, concepts, and methods ...        3 hours.   
...                                                 ...             ...   
3402  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3403  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3404  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3405  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3406    Approved for S/U grading only. May be repeated.  0 TO 16 hours.   

                                           Section Info  \
3026  Prerequisite: One of MATH 220 or MATH 221 or M...   
3027  Prerequisite: One of MATH 220 or MATH 221 or M...   
3028  Prerequisite: One of MATH 220 or MATH 221 or M...   
3029  Prerequisite: One of MATH 220 or MATH 221 or M...   
3030  Prerequisite: One of MATH 220 or MATH 221 or M...   
...                                                 ...   
3402  May be repeated in the same or separate terms ...   
3403  May be repeated in the same or separate terms ...   
3404  May be repeated in the same or separate terms ...   
3405  May be repeated in the same or separate terms ...   
3406    Approved for S/U grading only. May be repeated.   

                      Degree Attributes  ... Section Status  \
3026  Quantitative Reasoning II course.  ...              A   
3027  Quantitative Reasoning II course.  ...              A   
3028  Quantitative Reasoning II course.  ...              A   
3029  Quantitative Reasoning II course.  ...              A   
3030  Quantitative Reasoning II course.  ...              A   
...                                 ...  ...            ...   
3402                                NaN  ...              A   
3403                                NaN  ...              A   
3404                                NaN  ...              A   
3405                                NaN  ...              A   
3406                                NaN  ...              A   

      Enrollment Status                   Type Type Code Start Time  End Time  \
3026  Open (Restricted)                Lecture       LEC   11:00 AM  11:50 AM   
3027             Closed  Laboratory-Discussion       LBD   02:00 PM  03:50 PM   
3028             Closed  Laboratory-Discussion       LBD   04:00 PM  05:50 PM   
3029             Closed  Laboratory-Discussion       LBD   09:00 AM  10:50 AM   
3030               Open  Laboratory-Discussion       LBD   11:00 AM  12:50 PM   
...                 ...                    ...       ...        ...       ...   
3402  Open (Restricted)     Lecture-Discussion        S4   02:00 PM  03:15 PM   
3403             C

### Puzzle 3.3

Select only the rows that have a `Subject` of `IS`.  Assign this to the variable name `df_IS` instead of the usual `df`.
- Make sure your result contains only information science courses!

In [10]:
df_IS = df[df.Subject == "IS"]
df_IS

Year    Term YearTerm Subject  Number  \
6090  2024  Spring  2024-sp      IS     100   
6091  2024  Spring  2024-sp      IS     101   
6092  2024  Spring  2024-sp      IS     101   
6093  2024  Spring  2024-sp      IS     101   
6094  2024  Spring  2024-sp      IS     101   
...    ...     ...      ...     ...     ...   
6346  2024  Spring  2024-sp      IS     597   
6347  2024  Spring  2024-sp      IS     597   
6348  2024  Spring  2024-sp      IS     597   
6349  2024  Spring  2024-sp      IS     597   
6350  2024  Spring  2024-sp      IS     599   

                                                   Name  \
6090   Exploring the iSchool with a Human-Centered Lens   
6091               Introduction to Information Sciences   
6092               Introduction to Information Sciences   
6093               Introduction to Information Sciences   
6094               Introduction to Information Sciences   
...                                                 ...   
6346  Advanced Topics in Data Analytics &amp; Data S...   
6347  Advanced Topics in Data Analytics &amp; Data S...   
6348  Advanced Topics in Data Analytics &amp; Data S...   
6349  Advanced Topics in Data Analytics &amp; Data S...   
6350                                    Thesis Research   

                                            Description    Credit Hours  \
6090  This course introduces students to the School ...        1 hours.   
6091  This course provides an introduction to the fi...        3 hours.   
6092  This course provides an introduction to the fi...        3 hours.   
6093  This course provides an introduction to the fi...        3 hours.   
6094  This course provides an introduction to the fi...        3 hours.   
...                                                 ...             ...   
6346  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6347  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6348  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6349  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6350  Individual study and research. 0 to 16 graduat...  0 TO 16 hours.   

                                           Section Info  \
6090  Prerequisite: Restricted to Majors Only; First...   
6091                                                NaN   
6092                                                NaN   
6093                                                NaN   
6094                                                NaN   
...                                                 ...   
6346  2 or 4 graduate hours. No professional credit....   
6347  2 or 4 graduate hours. No professional credit....   
6348  2 or 4 graduate hours. No professional credit....   
6349  2 or 4 graduate hours. No professional credit....   
6350  0 to 16 graduate hours. No professional credit...   

                       Degree Attributes  ... Section Status  \
6090                                 NaN  ...              A   
6091  Social & Beh Sci - Soc Sci course.  ...              A   
6092  Social & Beh Sci - Soc Sci course.  ...              A   
6093  Social & Beh Sci - Soc Sci course.  ...              A   
6094  Social & Beh Sci - Soc Sci course.  ...              A   
...                                  ...  ...            ...   
6346                                 NaN  ...              A   
6347                                 NaN  ...              A   
6348                                 NaN  ...              A   
6349                                 NaN  ...              A   
6350                                 NaN  ...              A   

      Enrollment Status                   Type Type Code Start Time  End Time  \
6090  Open (Restricted)     Lecture-Discussion       LCD   04:00 PM  04:50 PM   
6091               Open  Discussion/Recitation       DIS   11:00 AM  12:20 PM   
6092               Open  Discussion/Recitation       DIS   11:00 AM  12:20 PM   
6093               Open  Disc

### 🔬 Test Case Checkpoint 🔬


In [11]:
## == CHECKPOINT TEST CASES for Part 3 ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.
assert( 'df_CS' in vars() ), "Make certain to name the CS courses df_CS."
assert( 'df_STAT' in vars() ), "Make certain to name the STAT courses df_STAT."
assert( 'df_IS' in vars() ), "Make certain to name the IS courses df_IS."
assert( len(df_STAT[df_STAT.Subject != "STAT"] ) == 0 ), "It looks like you did not subset df_STAT to only STAT courses."
assert( len(df_CS[df_CS.Subject != "CS"] ) == 0 ), "It looks like you did not subset df_CS to only CS courses."
assert( len(df_IS[df_IS.Subject != "IS"] ) == 0 ), "It looks like you did not subset df_IS to only IS courses."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


### Puzzle 3.4

Looking at the output of the DataFrames for STAT, CS, and IS above, which subject offers the **fewest** course sections?  The **most** course sections?  **BE SPECIFIC and mention which subject has the fewest/most sections.**


The subject that offers the fewest course sections is STAT since there are only 149 rows which means 149 course sections. The subject that offers the most course sections is CS since there are 381 rows which means 381 course sections. 

### Puzzle 3.5

Looking at the output of the DataFrames for STAT, CS, and IS above, which subject has the highest numbered course?  Is there something similar about all of the highest numbered courses?  Write about what you observed!


The highest course number is 599. The course is "Thesis Research" for STAT, CS, and IS. 

<hr style="color: #DD3403;">

# Part 4: Your Major

In this part, you will be exploring **your major** using the Course Catalog dataset.

- Use the primary subject of your **current major** (or, if you're undeclared, your intended major) for this section.
- Your answers will be different than the rest of your group!

### Puzzle 4.1:

Select only the rows that have a `Subject` of your current/intended major.  Assign this to the variable name `mydf` instead of `df`:

In [12]:
mydf = df[df.Subject == "ASRM"]
mydf

Year    Term YearTerm Subject  Number  \
1199  2024  Spring  2024-sp    ASRM     210   
1200  2024  Spring  2024-sp    ASRM     210   
1201  2024  Spring  2024-sp    ASRM     210   
1202  2024  Spring  2024-sp    ASRM     392   
1203  2024  Spring  2024-sp    ASRM     398   
1204  2024  Spring  2024-sp    ASRM     401   
1205  2024  Spring  2024-sp    ASRM     401   
1206  2024  Spring  2024-sp    ASRM     401   
1207  2024  Spring  2024-sp    ASRM     406   
1208  2024  Spring  2024-sp    ASRM     406   
1209  2024  Spring  2024-sp    ASRM     409   
1210  2024  Spring  2024-sp    ASRM     409   
1211  2024  Spring  2024-sp    ASRM     441   
1212  2024  Spring  2024-sp    ASRM     441   
1213  2024  Spring  2024-sp    ASRM     441   
1214  2024  Spring  2024-sp    ASRM     441   
1215  2024  Spring  2024-sp    ASRM     442   
1216  2024  Spring  2024-sp    ASRM     442   
1217  2024  Spring  2024-sp    ASRM     442   
1218  2024  Spring  2024-sp    ASRM     442   
1219  2024  Spring  2024-sp    ASRM     450   
1220  2024  Spring  2024-sp    ASRM     450   
1221  2024  Spring  2024-sp    ASRM     451   
1222  2024  Spring  2024-sp    ASRM     451   
1223  2024  Spring  2024-sp    ASRM     451   
1224  2024  Spring  2024-sp    ASRM     451   
1225  2024  Spring  2024-sp    ASRM     453   
1226  2024  Spring  2024-sp    ASRM     453   
1227  2024  Spring  2024-sp    ASRM     461   
1228  2024  Spring  2024-sp    ASRM     461   
1229  2024  Spring  2024-sp    ASRM     461   
1230  2024  Spring  2024-sp    ASRM     461   
1231  2024  Spring  2024-sp    ASRM     469   
1232  2024  Spring  2024-sp    ASRM     471   
1233  2024  Spring  2024-sp    ASRM     471   
1234  2024  Spring  2024-sp    ASRM     472   
1235  2024  Spring  2024-sp    ASRM     472   
1236  2024  Spring  2024-sp    ASRM     539   
1237  2024  Spring  2024-sp    ASRM     551   
1238  2024  Spring  2024-sp    ASRM     555   
1239  2024  Spring  2024-sp    ASRM     575   
1240  2024  Spring  2024-sp    ASRM     575   
1241  2024  Spring  2024-sp    ASRM     593   
1242  2024  Spring  2024-sp    ASRM     598   

                                                Name  \
1199                              Theory of Interest   
1200                              Theory of Interest   
1201                              Theory of Interest   
1202                       Actuarial Problem Solving   
1203                            Actuarial Internship   
1204                          Actuarial Statistics I   
1205                          Actuarial Statistics I   
1206                          Actuarial Statistics I   
1207      Linear Algebra with Financial Applications   
1208      Linear Algebra with Financial Applications   
1209  Stochastic Processes for Finance and Insurance   
1210  Stochastic Processes for Finance and Insurance   
1211                  Statistics for Risk Modeling I   
1212                  Statistics for Risk Modeling I   
1213                  Statistics for Risk Modeling I   
1214                  Statistics for Risk Modeling I   
1215                 Statistics for Risk Modeling II   
1216                 Statistics for Risk Modeling II   
1217                 Statistics for Risk Modeling II   
1218                 Statistics for Risk Modeling II   
1219                   Methods of Applied Statistics   
1220                   Methods of Applied Statistics   
1221                  Basics of Statistical Learning   
1222                  Basics of Statistical Learning   
1223                  Basics of Statistical Learning   
1224                  Basics of Statistical Learning   
1225                       Applied Bayesian Analysis   
1226                       Applied Bayesian Analysis   
1227                                     Loss Models   
1228                                     Loss Models   
1229                                     Loss Models   
1230                                     Loss Models   
1231                  Casualty Actuarial M

### Puzzle 4.2:

Select **five random course sections** from your major, using `mydf` and pandas functions:

In [14]:
mydf.sample(n = 5)

Year    Term YearTerm Subject  Number                            Name  \
1199  2024  Spring  2024-sp    ASRM     210              Theory of Interest   
1230  2024  Spring  2024-sp    ASRM     461                     Loss Models   
1211  2024  Spring  2024-sp    ASRM     441  Statistics for Risk Modeling I   
1231  2024  Spring  2024-sp    ASRM     469  Casualty Actuarial Mathematics   
1203  2024  Spring  2024-sp    ASRM     398            Actuarial Internship   

                                            Description   Credit Hours  \
1199  Study of compound interest and annuities; appl...       3 hours.   
1230  Foundation in the actuarial modeling process; ...       4 hours.   
1211  Introduction to statistics for risk modeling, ...       4 hours.   
1231  An introduction to property/casualty actuarial...  3 OR 4 hours.   
1203  Full-time or part-time practice of actuarial s...       0 hours.   

                                           Section Info Degree Attributes  \
1199              Prerequisite: MATH 231 or equivalent.               NaN   
1230  4 undergraduate hours. 4 graduate hours. Credi...               NaN   
1211  4 undergraduate hours. 4 graduate hours. Credi...               NaN   
1231  3 or 4 undergraduate hours. No graduate credit...               NaN   
1203  Approved for S/U grading only. May be repeated...               NaN   

      ... Section Status           Enrollment Status                   Type  \
1199  ...              A                        Open  Discussion/Recitation   
1230  ...              A                      Closed                Lecture   
1211  ...              A  CrossListOpen (Restricted)  Discussion/Recitation   
1231  ...              A           Open (Restricted)     Lecture-Discussion   
1203  ...              A           Open (Restricted)             Internship   

     Type Code Start Time  End Time Days of Week  Room  \
1199       DIS   08:00 AM  08:50 AM            F  1057   
1230       LEC   03:30 PM  04:50 PM           TR   259   
1211       DIS   10:00 AM  10:50 AM            F  4039   
1231       LCD   02:00 PM  03:20 PM           TR   106   
1203       INT   ARRANGED       NaN          NaN   NaN   

                           Building          Instructors  
1199                   Lincoln Hall  Wongwoottisaroch, Y  
1230               English Building             Icaza, E  
1211  Campus Instructional Facility             Icaza, E  
1231              David Kinley Hall            Freiji, C  
1203                            NaN            Freiji, C  

[5 rows x 27 columns]

### Puzzle 4.3:

Select **the seven lowest numbered course sections** from your major:

In [15]:
mydf.nsmallest(7, "Number")

Year    Term YearTerm Subject  Number                       Name  \
1199  2024  Spring  2024-sp    ASRM     210         Theory of Interest   
1200  2024  Spring  2024-sp    ASRM     210         Theory of Interest   
1201  2024  Spring  2024-sp    ASRM     210         Theory of Interest   
1202  2024  Spring  2024-sp    ASRM     392  Actuarial Problem Solving   
1203  2024  Spring  2024-sp    ASRM     398       Actuarial Internship   
1204  2024  Spring  2024-sp    ASRM     401     Actuarial Statistics I   
1205  2024  Spring  2024-sp    ASRM     401     Actuarial Statistics I   

                                            Description Credit Hours  \
1199  Study of compound interest and annuities; appl...     3 hours.   
1200  Study of compound interest and annuities; appl...     3 hours.   
1201  Study of compound interest and annuities; appl...     3 hours.   
1202  Methods and techniques of solving problems in ...     1 hours.   
1203  Full-time or part-time practice of actuarial s...     0 hours.   
1204                    Same as STAT 408. See STAT 408.     4 hours.   
1205                    Same as STAT 408. See STAT 408.     4 hours.   

                                           Section Info Degree Attributes  \
1199              Prerequisite: MATH 231 or equivalent.               NaN   
1200              Prerequisite: MATH 231 or equivalent.               NaN   
1201              Prerequisite: MATH 231 or equivalent.               NaN   
1202  Approved for S/U grading only. May be repeated...               NaN   
1203  Approved for S/U grading only. May be repeated...               NaN   
1204                    Same as STAT 408. See STAT 408.               NaN   
1205                    Same as STAT 408. See STAT 408.               NaN   

      ... Section Status  Enrollment Status                   Type Type Code  \
1199  ...              A               Open  Discussion/Recitation       DIS   
1200  ...              A             Closed  Discussion/Recitation       DIS   
1201  ...              A  Open (Restricted)                Lecture       LEC   
1202  ...              A  Open (Restricted)                Seminar       SEM   
1203  ...              A  Open (Restricted)             Internship       INT   
1204  ...              A      CrossListOpen  Discussion/Recitation       DIS   
1205  ...              A      CrossListOpen  Discussion/Recitation       DIS   

     Start Time  End Time Days of Week  Room                     Building  \
1199   08:00 AM  08:50 AM            F  1057                 Lincoln Hall   
1200   03:00 PM  03:50 PM            F  1024                 Lincoln Hall   
1201   09:30 AM  10:50 AM           TR   119  Materials Science & Eng Bld   
1202   05:00 PM  06:50 PM            M   313                 Gregory Hall   
1203   ARRANGED       NaN          NaN   NaN                          NaN   
1204   02:00 PM  02:50 PM            T  1060                 Lincoln Hall   
1205   03:00 PM  03:50 PM            T  1060                 Lincoln Hall   

              Instructors  
1199  Wongwoottisaroch, Y  
1200  Wongwoottisaroch, Y  
1201               Liu, P  
1202            Freiji, C  
1203            Freiji, C  
1204                  NaN  
1205                  NaN  

[7 rows x 27 columns]

### Puzzle 4.4:

Select **the nine highest numbered course sections** from your major:

In [16]:
mydf.nlargest(9, "Number")

Year    Term YearTerm Subject  Number  \
1242  2024  Spring  2024-sp    ASRM     598   
1241  2024  Spring  2024-sp    ASRM     593   
1239  2024  Spring  2024-sp    ASRM     575   
1240  2024  Spring  2024-sp    ASRM     575   
1238  2024  Spring  2024-sp    ASRM     555   
1237  2024  Spring  2024-sp    ASRM     551   
1236  2024  Spring  2024-sp    ASRM     539   
1234  2024  Spring  2024-sp    ASRM     472   
1235  2024  Spring  2024-sp    ASRM     472   

                                        Name  \
1242                      Literature Seminar   
1241                     Graduate Internship   
1239  Life Insurance and Pension Mathematics   
1240  Life Insurance and Pension Mathematics   
1238           Advanced Predictive Analytics   
1237                    Statistical Learning   
1236      Risk Analytics and Decision Making   
1234                   Life Contingencies II   
1235                   Life Contingencies II   

                                            Description   Credit Hours  \
1242  Students present seminars and discussions on a...  0 TO 4 hours.   
1241  Full-time or part-time practice of graduate-le...       0 hours.   
1239  Tabular and parametric survival models with si...       4 hours.   
1240  Tabular and parametric survival models with si...       4 hours.   
1238  Emphasizes applications of advanced predictive...       4 hours.   
1237        Same as CSE 542 and STAT 542. See STAT 542.       4 hours.   
1236  The course will give students the opportunity ...       2 hours.   
1234  Continuation of ASRM 471. Introduction to tabu...       3 hours.   
1235  Continuation of ASRM 471. Introduction to tabu...       3 hours.   

                                           Section Info Degree Attributes  \
1242  0 to 4 graduate hours. No professional credit....               NaN   
1241  0 graduate hours. No professional credit. Appr...               NaN   
1239  4 graduate hours. No professional credit. Cred...               NaN   
1240  4 graduate hours. No professional credit. Cred...               NaN   
1238  Same as STAT 541. 4 graduate hours. No profess...               NaN   
1237        Same as CSE 542 and STAT 542. See STAT 542.               NaN   
1236  2 graduate hours. No professional credit. Prer...               NaN   
1234  3 undergraduate hours. No graduate credit. Cre...               NaN   
1235  3 undergraduate hours. No graduate credit. Cre...               NaN   

      ... Section Status           Enrollment Status                   Type  \
1242  ...              A           Open (Restricted)     Lecture-Discussion   
1241  ...              A           Open (Restricted)             Internship   
1239  ...              A               CrossListOpen  Discussion/Recitation   
1240  ...              A  CrossListOpen (Restricted)                Lecture   
1238  ...              A  CrossListOpen (Restricted)     Lecture-Discussion   
1237  ...              A                      Closed     Lecture-Discussion   
1236  ...              A           Open (Restricted)     Lecture-Discussion   
1234  ...              A                      Closed  Discussion/Recitation   
1235  ...              A                      Closed                Lecture   

     Type Code Start Time  End Time Days of Week   Room           Building  \
1242       LCD   02:00 PM  02:50 PM            F  106B1   Engineering Hall   
1241       INT   ARRANGED       NaN          NaN    NaN                NaN   
1239       DIS   01:00 PM  01:50 PM            F   1022       Lincoln Hall   
1240       LEC   02:00 PM  03:20 PM           TR  106B1   Engineering Hall   
1238       LCD   12:30 PM  01:50 PM           TR    106  David Kinley Hall   
1237       LCD   03:00 PM  03:50 PM          MWF   1002       Lincoln Hall   
1236       LCD   09:00 AM  10:20 AM           MW    104   English Building   
1234       DIS   01:00 PM  01:50 PM            F   1022       Lincoln Hall   
1235       LEC   02:00 PM  03:20 PM           TR 

### Puzzle 4.5:

Select **a random sample of 10% of the course sections** from your major:

In [17]:
mydf.sample(frac = 0.1)

Year    Term YearTerm Subject  Number  \
1212  2024  Spring  2024-sp    ASRM     441   
1225  2024  Spring  2024-sp    ASRM     453   
1210  2024  Spring  2024-sp    ASRM     409   
1233  2024  Spring  2024-sp    ASRM     471   

                                                Name  \
1212                  Statistics for Risk Modeling I   
1225                       Applied Bayesian Analysis   
1210  Stochastic Processes for Finance and Insurance   
1233                            Life Contingencies I   

                                            Description   Credit Hours  \
1212  Introduction to statistics for risk modeling, ...       4 hours.   
1225                    Same as STAT 431. See STAT 431.  3 OR 4 hours.   
1210  An introduction to stochastic processes and th...  3 OR 4 hours.   
1233  Distribution of the time-to-death random varia...       4 hours.   

                                           Section Info Degree Attributes  \
1212  4 undergraduate hours. 4 graduate hours. Credi...               NaN   
1225                    Same as STAT 431. See STAT 431.               NaN   
1210  3 or 4 undergraduate hours. 3 or 4 graduate ho...               NaN   
1233  4 undergraduate hours. 4 graduate hours. Prere...               NaN   

      ... Section Status           Enrollment Status                Type  \
1212  ...              A  CrossListOpen (Restricted)             Lecture   
1225  ...              A                      Closed  Lecture-Discussion   
1210  ...              A  CrossListOpen (Restricted)  Lecture-Discussion   
1233  ...              A           Open (Restricted)             Lecture   

     Type Code Start Time  End Time Days of Week   Room           Building  \
1212       LEC   11:00 AM  12:20 PM           TR    119  David Kinley Hall   
1225       LCD   02:00 PM  02:50 PM          MWF    319       Gregory Hall   
1210       LCD   03:00 PM  04:20 PM           MW  106B1   Engineering Hall   
1233       LEC   09:30 AM  10:50 AM           TR    137             Armory   

     Instructors  
1212    Icaza, E  
1225     Park, T  
1210   Fadina, T  
1233      Wei, W  

[4 rows x 27 columns]

### ✔️ Checkpoint ✔️

There are no test cases for Part 4 (since everyone has different majors), but go double check your work!  There's just one section left to go! :)

<hr style="color: #DD3403;">

# Part 5: Growth of the IS major

The IS major is still a new relatively new major here at UIUC. The final puzzle is for you to analyze how much the IS major has grown since 2019.

### Puzzle 5.1: Loading the Course Catalog Dataset from 2019

Using the name `df2019` instead of `df`, load the "Fall 2019 Course Catalog Dataset" that can be found at the following URL:
```
https://waf.cs.illinois.edu/discovery/course-catalog-fa19.csv
```

In [23]:
df2019 = pd.read_csv("https://waf.cs.illinois.edu/discovery/course-catalog-fa19.csv")
df2019

Year  Term YearTerm Subject  Number  \
0     2019  Fall  2019-fa     AAS     100   
1     2019  Fall  2019-fa     AAS     105   
2     2019  Fall  2019-fa     AAS     120   
3     2019  Fall  2019-fa     AAS     199   
4     2019  Fall  2019-fa     AAS     200   
...    ...   ...      ...     ...     ...   
8584  2019  Fall  2019-fa    ZULU     202   
8585  2019  Fall  2019-fa    ZULU     403   
8586  2019  Fall  2019-fa    ZULU     404   
8587  2019  Fall  2019-fa    ZULU     405   
8588  2019  Fall  2019-fa    ZULU     406   

                                       Name  \
0              Intro Asian American Studies   
1     Introduction to Arab American Studies   
2             Intro to Asian Am Pop Culture   
3                Undergraduate Open Seminar   
4                      U.S. Race and Empire   
...                                     ...   
8584                     Elementary Zulu II   
8585                    Intermediate Zulu I   
8586                   Intermediate Zulu II   
8587                        Advanced Zulu I   
8588                       Advanced Zulu II   

                                            Description   Credit Hours  \
0     Interdisciplinary introduction to the basic co...       3 hours.   
1     Interdisciplinary introduction to the basic co...       3 hours.   
2     Introductory understanding of the way U.S. pop...       3 hours.   
3              May be repeated to a maximum of 6 hours.  1 TO 5 hours.   
4     Invites students to examine histories and narr...       3 hours.   
...                                                 ...            ...   
8584  Continuation of ZULU 201 with introduction of ...       5 hours.   
8585  Survey of more advanced grammar; emphasis on i...       4 hours.   
8586  Continuation of ZULU 403; emphasis on increasi...       4 hours.   
8587  Third year Zulu with emphasis on conversationa...       3 hours.   
8588  Continuation of Zulu 405 with increased emphas...       3 hours.   

                                           Section Info  \
0                                                   NaN   
1                                                   NaN   
2                                                   NaN   
3                                                   NaN   
4                                      Same as LLS 200.   
...                                                 ...   
8584  Same as AFST 252. Participation in the languag...   
8585                                                NaN   
8586                                                NaN   
8587                                                NaN   
8588                                                NaN   

                                      Degree Attributes  
0     Social & Beh Sci - Soc Sci, and Cultural Studi...  
1                Cultural Studies - US Minority course.  
2                Cultural Studies - US Minority course.  
3                                                   NaN  
4     Humanities - Hist & Phil, and Cultural Studies...  
...                                                 ...  
8584                                                NaN  
8585                                                NaN  
8586                                                NaN  
8587                                                NaN  
8588                                                NaN  

[8589 rows x 10 columns]

### Puzzle 5.2: Data Science Analysis

In the following cells,
- Select only the rows from the `IS` subject from the `df2019` dataset,
- Select only the rows from the `IS` subject from the `df` dataset (this contains the Spring 2024 data),
- Complete the reflection question below (Puzzle 5.3).  Feel free to write more code to answer it.

In [24]:
x = df2019[df2019.Subject == "IS"]
x

Year  Term YearTerm Subject  Number  \
4843  2019  Fall  2019-fa      IS     107   
4844  2019  Fall  2019-fa      IS     199   
4845  2019  Fall  2019-fa      IS     202   
4846  2019  Fall  2019-fa      IS     310   
4847  2019  Fall  2019-fa      IS     351   
...    ...   ...      ...     ...     ...   
4930  2019  Fall  2019-fa      IS     591   
4931  2019  Fall  2019-fa      IS     592   
4932  2019  Fall  2019-fa      IS     593   
4933  2019  Fall  2019-fa      IS     594   
4934  2019  Fall  2019-fa      IS     599   

                                             Name  \
4843                       Data Science Discovery   
4844                   Undergraduate Open Seminar   
4845                     Social Aspects Info Tech   
4846                  Computing in the Humanities   
4847  The Design of Usable Information Interfaces   
...                                           ...   
4930                                    Practicum   
4931                            Independent Study   
4932                                  CAS Project   
4933                Information Sciences Practice   
4934                              Thesis Research   

                                            Description    Credit Hours  \
4843          Same as CS 107 and STAT 107. See STAT 107        4 hours.   
4844       Undergraduate Open Seminar. May be repeated.   1 TO 5 hours.   
4845       Same as INFO 202 and MACS 202. See INFO 202.        3 hours.   
4846  Explores use and application of technology to ...        3 hours.   
4847  Examines issues of Human Computer Interaction ...        3 hours.   
...                                                 ...             ...   
4930  Supervised field experience of professional-le...        2 hours.   
4931  Permits the intermediate or advanced student o...   2 TO 4 hours.   
4932  Individual study of a problem in library and i...   0 TO 8 hours.   
4933  Full-time or part-time practice of of any info...        0 hours.   
4934  Individual study and research. 0 to 16 graduat...  0 TO 16 hours.   

                                           Section Info  \
4843          Same as CS 107 and STAT 107. See STAT 107   
4844                                                NaN   
4845       Same as INFO 202 and MACS 202. See INFO 202.   
4846                                                NaN   
4847                  Prerequisite: Sophomore standing.   
...                                                 ...   
4930  2 graduate hours. No professional credit. Appr...   
4931  2 to 4 graduate hours. No professional credit....   
4932  0 to 8 graduate hours. No professional credit....   
4933  0 graduate hours. No professional credit. Appr...   
4934  0 to 16 graduate hours. No professional credit...   

                       Degree Attributes  
4843    Quantitative Reasoning I course.  
4844                                 NaN  
4845  Social & Beh Sci - Soc Sci course.  
4846                                 NaN  
4847                                 NaN  
...                                  ...  
4930                                 NaN  
4931                                 NaN  
4932                                 NaN  
4933                                 NaN  
4934                                 NaN  

[92 rows x 10 columns]

In [25]:
y = df[df.Subject == "IS"]
y

Year    Term YearTerm Subject  Number  \
6090  2024  Spring  2024-sp      IS     100   
6091  2024  Spring  2024-sp      IS     101   
6092  2024  Spring  2024-sp      IS     101   
6093  2024  Spring  2024-sp      IS     101   
6094  2024  Spring  2024-sp      IS     101   
...    ...     ...      ...     ...     ...   
6346  2024  Spring  2024-sp      IS     597   
6347  2024  Spring  2024-sp      IS     597   
6348  2024  Spring  2024-sp      IS     597   
6349  2024  Spring  2024-sp      IS     597   
6350  2024  Spring  2024-sp      IS     599   

                                                   Name  \
6090   Exploring the iSchool with a Human-Centered Lens   
6091               Introduction to Information Sciences   
6092               Introduction to Information Sciences   
6093               Introduction to Information Sciences   
6094               Introduction to Information Sciences   
...                                                 ...   
6346  Advanced Topics in Data Analytics &amp; Data S...   
6347  Advanced Topics in Data Analytics &amp; Data S...   
6348  Advanced Topics in Data Analytics &amp; Data S...   
6349  Advanced Topics in Data Analytics &amp; Data S...   
6350                                    Thesis Research   

                                            Description    Credit Hours  \
6090  This course introduces students to the School ...        1 hours.   
6091  This course provides an introduction to the fi...        3 hours.   
6092  This course provides an introduction to the fi...        3 hours.   
6093  This course provides an introduction to the fi...        3 hours.   
6094  This course provides an introduction to the fi...        3 hours.   
...                                                 ...             ...   
6346  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6347  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6348  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6349  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6350  Individual study and research. 0 to 16 graduat...  0 TO 16 hours.   

                                           Section Info  \
6090  Prerequisite: Restricted to Majors Only; First...   
6091                                                NaN   
6092                                                NaN   
6093                                                NaN   
6094                                                NaN   
...                                                 ...   
6346  2 or 4 graduate hours. No professional credit....   
6347  2 or 4 graduate hours. No professional credit....   
6348  2 or 4 graduate hours. No professional credit....   
6349  2 or 4 graduate hours. No professional credit....   
6350  0 to 16 graduate hours. No professional credit...   

                       Degree Attributes  ... Section Status  \
6090                                 NaN  ...              A   
6091  Social & Beh Sci - Soc Sci course.  ...              A   
6092  Social & Beh Sci - Soc Sci course.  ...              A   
6093  Social & Beh Sci - Soc Sci course.  ...              A   
6094  Social & Beh Sci - Soc Sci course.  ...              A   
...                                  ...  ...            ...   
6346                                 NaN  ...              A   
6347                                 NaN  ...              A   
6348                                 NaN  ...              A   
6349                                 NaN  ...              A   
6350                                 NaN  ...              A   

      Enrollment Status                   Type Type Code Start Time  End Time  \
6090  Open (Restricted)     Lecture-Discussion       LCD   04:00 PM  04:50 PM   
6091               Open  Discussion/Recitation       DIS   11:00 AM  12:20 PM   
6092               Open  Discussion/Recitation       DIS   11:00 AM  12:20 PM   
6093               Open  Disc

In [ ]:
# Feel free to add more code cells if needed by hovering over the area below the cell and pressing "+ Code"!

### Puzzle 5.3: Reflection on the Growth of Information Science

In the follow cell, write **at least one complete sentence explaining what you found** related to the growth of the IS program.  In your answer, **use actual data** (ex: by how much did it grow?)

In [ ]:
In 2019, there were only 92 course sections whereas in 2024, there are now 261 sections. This shows an increase of 169 courses offered from 2019-2024.

<hr style="color: #DD3403;">

# Submission

You're almost done!  All you need to do is to commit your lab to GitHub:

1.  ⚠️ **Make certain to save your work.** ⚠️ To do this, go to **File => Save All**

2.  After you have saved, exit this notebook and follow the Canvas instructions to commit this lab to your Git repository!

3. Your TA will grade your submission and provide you feedback after the lab is due. :)